Add kernel.org's `.clang-format` for editor-agnostic C formatting hints

bleggett opened the pull request:
`clang-format` is an editor-agnostic means of describing C formatting prefs for code editors.

Since upstream kernel uses this to define an explicit contract for C formatting, and our
BPF code follows the kernel style, this just brings in a copy of the kernel `.clang-format` dotfile.

This will get picked up by any editor that supports it (VSCode w/ C extension, emacs with `lsp`, vims, etc).

This will declare a specific contract for contributors around C style, as well as help this project explicitly declare any deviations from that style.

This is a straight copy of the [kernel.org file](https://github.com/torvalds/linux/blob/master/.clang-format), with a kernel-specific macro allowlist removed.

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+IncludeIsMainRegex: '(Test)?$'
+IndentCaseLabels: false
+IndentGotoLabels: false
+IndentPPDirectives: None
```
We could use this:

```suggestion
IndentPPDirectives: AfterHash
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#indentppdirectives

Although this indents directives everywhere, including where we're not currently using them. Meaning that if, for example, we remove one level of nested `#ifdef`s, we have to update the indent for inner directives.

If using the above we'd also need to set `PPIndentWidth`, because it seems to default to `IndentWidth` but we don't want that:

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#ppindentwidth

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+IndentCaseLabels: false
+IndentGotoLabels: false
+IndentPPDirectives: None
+IndentWidth: 8
```
```suggestion
IndentWidth: 8
InsertNewlineAtEOF: true
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#insertnewlineateof

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+PenaltyExcessCharacter: 100
+PenaltyReturnTypeOnItsOwnLine: 60
+
+PointerAlignment: Right
```
```suggestion
PointerAlignment: Right
QualifierAlignment: Left
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#qualifieralignment

Note: There's an option to specify the list for sorting the different qualifiers, but I haven't been able to get it to work on clang-format-17.

bleggett replied:
I'm inclined to leave this one as `Leave` since it apparently can break things/change intent if set to anything else.

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+SpaceBeforeCtorInitializerColon: true
+SpaceBeforeInheritanceColon: true
+SpaceBeforeParens: ControlStatements
+SpaceBeforeRangeBasedForLoopColon: true
```
Seems like we're leaving no space in `{}` most of the time, we could as well enforce it.

```suggestion
SpaceBeforeRangeBasedForLoopColon: true
SpaceInEmptyBlock: true
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#spaceinemptyblock

bleggett replied:
`checkpatch` has no opinions about this AFAICT, and the current codebase default seems to be `SpaceInEmptyBlock: false`.

When `checkpatch` doesn't have an opinion, it's probably best to go with a setting that reflects the existing style already in the codebase (for now, anyway) - I can't find any existing instances of `{ }` on a cursory look.

qmonnet replied:
Apologies, I meant to set it to `false`, but messed up when writing the comment.

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+AlignConsecutiveDeclarations: false
+AlignEscapedNewlines: Left
+AlignOperands: true
+AlignTrailingComments: false
```
Looks like we might to change this one, we're aligning comments at multiple locations

```suggestion
AlignTrailingComments: Always
```

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+---
+AccessModifierOffset: -4
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: false
+AlignConsecutiveDeclarations: false
```
Note: We would like to enable these inside of struct/enum/macro declarations and definitions, but it looks like it's not possible :disappointed:. Looks like it's either aligned or not, for everywhere.

bleggett replied:
For now, suggest leaving it as `false` - yep.

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+AccessModifierOffset: -4
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: false
+AlignConsecutiveDeclarations: false
```
We can at least enable for macros, though:

```suggestion
AlignConsecutiveDeclarations: false
AlignConsecutiveMacros:
  Enabled: true
  AcrossEmptyLines: true
  AcrossComments: true
```

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignconsecutivemacros

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+AlwaysBreakBeforeMultilineStrings: false
+AlwaysBreakTemplateDeclarations: false
+BinPackArguments: true
+BinPackParameters: true
```
```suggestion
BinPackParameters: true
BitFieldColonSpacing: None
```

`clang-format` also seems to move bitfields components onto a single line, but I've not found how to disable that :disappointed:.

We're also getting spaces before colons for labels, but I'm not sure how to disable that, either.

bleggett replied:
Can you point me at a file where you see this?

qmonnet replied:
- For bitfields: look for `[^ ]:[^ ]`, first example I see is in `struct policy_entry` in `bpf/lib/common.h`.

- For the space between label and colon: `skip_egress_gateway` in `bpf/bpf_lxc.c` is the first example I could find.

bleggett replied:
Bitfields: Yeah looks like we either live with this, or we avoid multiple bitfield declarations in a single statement, e.g. instead of:

```
__u8		deny:1,
		wildcard_protocol:1, /* protocol is fully wildcarded */
		wildcard_dport:1, /* dport is fully wildcarded */
```

we go with 

```
__u8		deny:1;
__u8		wildcard_protocol:1; /* protocol is fully wildcarded */
__u8		wildcard_dport:1; /* dport is fully wildcarded */
```

Either way, clang doesn't have a way to handle both and `checkpatch` doesn't care, so for now it has to be manually reconciled.

bleggett replied:
Label space:  that seems to be controlled by `SpaceBeforeInheritanceColon` - which seems like a bug, but setting `SpaceBeforeInheritanceColon: false` is probably fine for us given that we don't (afaik) use any class inheritance in bpf code. 

I'm going to set `SpaceBeforeInheritanceColon: false` and add a comment indicating this quirk.

qmonnet commented on .clang-format:
```diff
@@ -0,0 +1,120 @@
+PenaltyBreakFirstLessLess: 0
+PenaltyBreakString: 10
+PenaltyExcessCharacter: 100
+PenaltyReturnTypeOnItsOwnLine: 60
```
Looks like we need to experiment with this value, but I haven't been able to understand how it works, yet.

bleggett replied:
I added some comments around how PenaltyXXX settings work.

I tuned them a _bit_, mostly to get `PenaltyReturnTypeOnItsOwnLine` to stick to existing code style - feels a bit better now but we can twiddle these weights more.

qmonnet reviewed (commented):
Various comments and observations. I experimented a little, not something exhaustive.

(Additions or updates to the values we got from the kernel's file, if any, should likely go into a separate commit.)

bleggett committed:
Update CODEOWNERS as requested for `.clang-format`

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

bleggett committed:
Bring in .clang-format from torvalds/linux

https://github.com/torvalds/linux/blob/7acc1372113083fa281ba426021801e2402caca1/.clang-format

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

bleggett committed:
Apply initial cilium-specific tweaks

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

qmonnet reviewed (approved):
Looks all good, thanks a lot!

> please let me know about `IndentPPDirectives: AfterHash`

OK let's leave it out for now, or let's see if others chime in. Hard to tell what option would be the best here anyway.

The checkpatch workflow reports that your commit subject is too long on your first commit, I think it's because you're just missing the empty line between the subject and the description (the link).

Note to other reviewers: This PR was discussed at the community meeting last Wednesday (2023-05-24) and the consensus was to add `.clang-format`, nobody raised any particular objections.
